synapse/storage/databases/main/relations.py

```diff
@@ -363,37 +378,67 @@ async def get_applicable_edit(
 
         # Fetches latest edit that has the same type and sender as the
         # original, and is an `m.room.message`.
-        sql = """
-            SELECT edit.event_id FROM events AS edit
-            INNER JOIN event_relations USING (event_id)
-            INNER JOIN events AS original ON
-                original.event_id = relates_to_id
-                AND edit.type = original.type
-                AND edit.sender = original.sender
-            WHERE
-                relates_to_id = ?
-                AND relation_type = ?
-                AND edit.room_id = ?
-                AND edit.type = 'm.room.message'
-            ORDER by edit.origin_server_ts DESC, edit.event_id DESC
-            LIMIT 1
-        """
+        if isinstance(self.database_engine, PostgresEngine):
+            # The `DISTINCT ON` clause will pick the *first* row it encounters,
+            # so ordering by origin server ts + event ID desc will ensure we get
+            # the latest edit.
```

Comment:
I'm not sure what the event ID has to do with it. Aren't they some pseudorandom identifier? Or perhaps the point is that we want a completely deterministic order in case of edits with the same timestamp?

Reply:
We need a deterministic way for all servers to come to the same conclusion.